src/components/implementation/no_interface/scheddev/sl_lock.c

```diff
@@ -0,0 +1,110 @@
+#include "sl_lock.h"
+
+#include "sl.h"
+
+void sl_lock_init(struct sl_lock *lock) {
+    assert(lock);
+
+    *lock = (struct sl_lock) {
+        .holder = 0
+    };
+}
+
+thdid_t sl_lock_holder(struct sl_lock *lock) {
+    assert(lock);
+    return lock->holder;
+}
+
+// If timeout is 0, ignore the timeout
+int sl_lock_timeout_elapsed(microsec_t start_time, microsec_t timeout) {
+    if (timeout == 0 || (sl_now_usec() - start_time) <= timeout) {
+        return 0;
+    }
+    return 1;
+}
+
+int sl_lock_lock_internal_no_cs(struct sl_lock *lock, microsec_t timeout) {
+    assert(lock);
+
+    microsec_t start_time = sl_now_usec();
```

Comment:
The only reason to use `start_time` and `delay` in the timing APIs is to avoid races that plague solely relative time APIs.  Not taking the start time from the user makes this point moot.  So I don't understand why `sl_lock_timeout_elapsed` exists.  If we don't care about that race (which we can talk about), then we should just calculate the deadline immediately (`microsec_t deadline = sl_now_usec() + timeout;`), and just compare that to now_us.